src/intervals/stream_buffer.jl

```diff
@@ -0,0 +1,94 @@
+
+
+@doc """
+A buffer in which pushing to the back, popping from the front (shift),
+and indexing is efficient.
+
+It's constructed as a circular buffer that resizes when it runs out of space.
+""" ->
+type StreamBuffer{T}
+    data::Vector{T}
+    first_idx::Int
+    last_idx::Int
+
+    function StreamBuffer()
+        return new(Array(T, 16), 1, 0)
+    end
+end
+
+
+function Base.length{T}(buf::StreamBuffer{T})
+    if buf.last_idx == 0
+        return 0
+    elseif buf.last_idx < buf.first_idx
+        return length(buf.data) + buf.last_idx - buf.first_idx + 1
+    else
+        return buf.last_idx - buf.first_idx + 1
+    end
+end
+
+
+function Base.isempty{T}(buf::StreamBuffer{T})
+    return buf.last_idx == 0
+end
+
+
+function Base.endof{T}(buf::StreamBuffer{T})
```

Comment:
This generic function should imported explicit on this module, therefore we wouldn't need `Base.` here.